compiler/rustc_resolve/src/late/diagnostics.rs

```diff
@@ -1682,6 +1696,167 @@ impl<'a: 'ast, 'ast, 'tcx> LateResolutionVisitor<'a, '_, 'ast, 'tcx> {
         true
     }
 
+    fn suggest_alternative_construction_methods(
```

Comment:
I'm not fond of having twice (almost) identical code in two parts of the compiler. Which version has the most effect? The two tests I see exercise the non-local version, so iiuc this one.

Reply:
Yeah, it's not ideal, but we have different amount of context on each stage.

Reply:
This one affects only the first cases in `tests/ui/privacy/suggest-box-new.rs`, while the other affects the latter cases in that file. We have no way of knowing what the prevalence of these errors are in the wild. I you'd prefer I moved the logic to a central place I can do that. I didn't because there was no natural place where these wouldn't feel out of place.